test/Microsoft.ML.Functional.Tests/IntrospectiveTraining.cs

```diff
@@ -0,0 +1,431 @@
+
+        /// <summary>
+        /// Introspective Training: Linear model parameters may be inspected.
+        /// </summary>
+        [Fact]
+        public void InpsectLinearModelParameters()
+        {
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            var data = mlContext.Data.LoadFromTextFile<TweetSentiment>(GetDataPath(TestDatasets.Sentiment.trainFilename),
+                hasHeader: TestDatasets.Sentiment.fileHasHeader,
+                separatorChar: TestDatasets.Sentiment.fileSeparator,
+                allowQuoting: TestDatasets.Sentiment.allowQuoting);
+
+            // Create a training pipeline.
+            var pipeline = mlContext.Transforms.Text.FeaturizeText("Features", "SentimentText")
+                .AppendCacheCheckpoint(mlContext)
+                .Append(mlContext.BinaryClassification.Trainers.StochasticDualCoordinateAscentNonCalibrated(
+                    new SdcaNonCalibratedBinaryTrainer.Options { NumThreads = 1 }));
+
+            // Fit the pipeline.
+            var model = pipeline.Fit(data);
+
+            // Transform the data.
+            var transformedData = model.Transform(data);
+
+            // Extract the linear model from the pipeline.
+            var linearModel = model.LastTransformer.Model;
+
+            // Get the model bias and weights.
+            var bias = linearModel.Bias;
+            var weights = linearModel.Weights;
+
+            // Make sure the model weights array is the same length as the features array.
+            var numFeatures = (transformedData.Schema["Features"].Type as VectorType).Size;
+            Assert.Equal(numFeatures, weights.Count);
+        }
+
+        /// <summary>
+        /// Introspectable Training: Parameters of a trained Normalizer are easily accessed.
+        /// </summary>
+        [Fact]
+        void IntrospectNormalization()
+        {
+            // Concurrency must be 1 to assure that the mapping is done sequentially.
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            // Load the Iris dataset.
+            var data = mlContext.Data.LoadFromTextFile<Iris>(
+                GetDataPath(TestDatasets.iris.trainFilename),
+                hasHeader: TestDatasets.iris.fileHasHeader,
+                separatorChar: TestDatasets.iris.fileSeparator);
+
+            // Compose the transformation.
+            var pipeline = mlContext.Transforms.Concatenate("Features", Iris.Features)
+                .Append(mlContext.Transforms.Normalize("Features", mode: NormalizingEstimator.NormalizerMode.MinMax));
+
+            // Fit the pipeline.
+            var model = pipeline.Fit(data);
+
+            // Extract the normalizer from the trained pipeline.
+            // TODO #2854: Extract the normalizer parameters.
+            var normalizer = model.LastTransformer;
+        }
+
+        /// <summary>
+        /// I can take an existing model file and inspect what transformers were included in the pipeline. 	 
```

Comment:
>I can take an existing model file [](start = 12, length = 33)

You are not taking a model file. You are constructing the pipeline in the test.

Reply:
Good point. I am updating the summary. I changed this test to just look at pipelines, and not necessarily at serialization / deserialization. There will be model-file-specific tests that test serialization and deserialization, so I decided to not test that here.

---
In reply to: [262709151](https://github.com/dotnet/machinelearning/pull/2859#discussion_r262709151) [](ancestors = 262709151)